src/qibo/models/grover.py

```diff
@@ -0,0 +1,194 @@
+
+        Returns:
+            measured (str): bitstring measured and checked as a valid solution.
+            total_iterations (int): number of times the oracle has been called.
+        """
+        k = 1
+        lamda = lamda_value
+        total_iterations = 0
+        while True:
+            it = np.random.randint(k + 1)
+            if it != 0:
+                total_iterations += it
+                circuit = self.circuit(it)
+                result = circuit(nshots=1)
+                measured = result.frequencies(binary=True).most_common(1)[0][0]
+                if self.check(measured, *self.check_args):
+                    return measured, total_iterations
+            k = min(lamda * k, np.sqrt(self.sup_size))
+            if total_iterations > (9/4) * np.sqrt(self.sup_size):
+                raise_error(TimeoutError, "Cancelling iterative method as too "
+                                          "many iterations have taken place.")
+
+    def execute(self, nshots=100, freq=False):
+        """Execute Grover's algorithm.
+
+        If the number of solutions is given, calculates iterations,
+        otherwise it uses an iterative approach.
+
+        Args:
+            nshots (int): number of shots in order to get the frequencies.
+            freq (bool): print the full frequencies after the exact Grover algorithm.
+
+        Returns:
+            solution (str): bitstring (or list of bitstrings) measured as solution of the search.
+            iterations (int): number of oracle calls done to reach a solution.
+        """
+        if self.num_sol and not self.iterative:
+            it = int(np.pi * np.sqrt(self.sup_size / self.num_sol) / 4)
+            circuit = self.circuit(it)
+            result = circuit(nshots=nshots).frequencies(binary=True)
+            if freq:
+                log.info("Result of sampling Grover's algorihm")
+                log.info(result)
+                self.frequencies = result
+            log.info(f"Most common states found using Grover's algorithm with {it} iterations:")
+            most_common = result.most_common(self.num_sol)
+            self.solution = []
+            self.iterations = it
+            for i in most_common:
+                log.info(i[0])
+                self.solution.append(i[0])
+                if self.check:
+                    if self.check(i[0], *self.check_args):
+                        log.info('Solution checked and successful.')
+                    else:
+                        log.info('Not a solution of the problem. Something went wrong.')
+        else:
+            if not self.check:
+                raise_error(ValueError, "Check function needed for iterative approach.")
+            measured, total_iterations = self.iterative_grover()
+            log.info('Solution found in an iterative process.')
+            log.info(f'Solution: {measured}')
+            log.info(f'Total Grover iterations taken: {total_iterations}')
```

Comment:
I would probably remove all printing from the model and make sure that all the relevant information related to the execution is returned instead of printed. This is more flexible as it allows the user to print or use this information in any way. Alternatively we can add a flag in the `execute` arguments that chooses whether messages are printed or no.

Reply:
I think adding a flag that is False by default might be the best compromise.